File: itoa-divless/itoa_divless_sse.py
#!/usr/bin/env python3

def main():
    with open('itoa_divless_sse_table.h', 'wt') as f:
        lookup(f, 'lookup0', shift=0)
        lookup(f, 'lookup1', shift=8)
        lookup(f, 'lookup2', shift=16)
        lookup(f, 'lookup3', shift=24)


def lookup(f, name, shift):
    f.write(f'static uint8_t {name}[256][16] = {{\n')
    for i, values in enumerate(lookup_values(shift)):
        f.write('{%s}' % ', '.join(values))
        if i != 255:
            f.write(',')
        f.write('\n')

    f.write('};\n')


def lookup_values(shift):
    values = []
    for b in range(256):
        v = b << shift
        t = list(f'{v:016d}')
        values.append(t)

    return values


if __name__ == '__main__':
    main()

File: itoa-divless/itoa_divless_sse.cpp
#include "itoa_divless_sse.h"

#include "itoa_divless_sse_table.h"

#include <immintrin.h>

void itoa_divless_sse(uint32_t x, char* buffer) {
    __m128i BCD = _mm_setzero_si128();

    // 1. obtain BCD representation of all bytes
    {
        const uint8_t byte0 = x & 0xff;
        BCD = _mm_add_epi8(BCD, _mm_loadu_si128((const __m128i*)&lookup0[byte0][0]));
    }
    {
        const uint8_t byte1 = (x >> 8) & 0xff;
        BCD = _mm_add_epi8(BCD, _mm_loadu_si128((const __m128i*)&lookup1[byte1][0]));
    }
    {
        const uint8_t byte2 = (x >> 16) & 0xff;
        BCD = _mm_add_epi8(BCD, _mm_loadu_si128((const __m128i*)&lookup2[byte2][0]));
    }
    {
        const uint8_t byte3 = (x >> 24) & 0xff;
        BCD = _mm_add_epi8(BCD, _mm_loadu_si128((const __m128i*)&lookup3[byte3][0]));
    }

    // 2. fixup BCD & store result
    const __m128i shift_right_1 = _mm_setr_epi8(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, -1);

    while (true) {
        const __m128i mask = _mm_cmpgt_epi8(BCD, _mm_set1_epi8(9));
        if (_mm_movemask_epi8(mask) == 0)
            break;

        // carry[i] = 0 or -1
        const __m128i carry = _mm_shuffle_epi8(mask, shift_right_1);

        // byte[i] -= 10 if byte[i] >= 10
        BCD = _mm_sub_epi8(BCD, _mm_and_si128(mask, _mm_set1_epi8(10)));

        // byte[i] += carry[i - 1]
        BCD = _mm_sub_epi8(BCD, carry);
    }

    const __m128i ascii = _mm_add_epi8(BCD, _mm_set1_epi8('0'));

    _mm_storeu_si128((__m128i*)buffer, ascii);
}
